File: UCF HSPT Documents/2007/Solutions/roboto.cpp

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Global variables
int junctions[512][512];
int numJunctions;

// A number big enough to represent infinity for this problem
#define INFINITY 10000


// To solve this problem, we use a wonderful algorithm known as 
// Floyd-Warshall's All Shortest Paths.  This lets us find the
// shortest path (the path of least resistance) from any junction to any
// other.  Remember that we're looking for short circuits in the design.
// Since a short circuit has a resistance of zero, it will always be the
// path of least resistance between any two junctions, and will always
// show up when we use Floyd-Warshall's algorithm.
//
// We start with the original circuit design matrix.  Each entry (i,j) in
// this matrix indicates the length (or resistance in this case) of the
// direct connection between junctions i and j.
// After running Floyd-Warshall's on this matrix, it will contain the
// length of the shortest path (or the path of least resistance in this
// case) between junctions i and j.  This path may still go directly
// from i to j, or it may go through several other junctions to get
// from i to j.
//
// Note that this algorithm is a bit of overkill for this problem, since
// it gives us the shortest path for every pair of junctions, and we only
// care about the path from junction 0 to junction 1.  However, 
// Floyd-Warshall's is the simplest shortest paths algorithm to implement
// in code (which makes it good to use in contest programming), and it works
// quite well, as long as the problem isn't too large (with only 500 junctions
// maximum, this problem is plenty small enough).
//
void floyd()
{
   int i, j, k;

   // We have to do a simple preprocessing step here.  The -1's in the
   // circuit design matrix will cause this algorithm to fail without a
   // lot of messy code to handle this special case. To fix this, we'll
   // convert the -1's to a large number.  You can think of this number as
   // representing infinite resistance, which is consistent with the lack
   // of a connection between two junctions.
   for (i = 0; i < numJunctions; i++)
      for (j = 0; j < numJunctions; j++)
         if (junctions[i][j] < 0)
            junctions[i][j] = INFINITY;

   // This outer loop (k) will test paths going from junction i to junction
   // j that go through junction k.
   for (k = 0; k < numJunctions; k++)
   {
      // The inner loops simply iterate over all junctions, testing for
      // less resistant paths from i to j
      for (i = 0; i < numJunctions; i++)
      {
         for (j = 0; j < numJunctions; j++)
         {
            // If there is a path from i to j going through k that is
            // less resistant than any path we've found so far, remember
            // the resistance along this new path
            if (junctions[i][k] + junctions[k][j] < junctions[i][j])
               junctions[i][j] = junctions[i][k] + junctions[k][j];
         }
      }
   }
}


int main(void)
{
   FILE *fp;
   char line[256];
   char *token;
   int  numDesigns;
   int  designNum;
   int  i, j;

   // Open the input file
   fp = fopen("roboto.in", "r");

   // Read the number of designs in the input
   fgets(line, sizeof(line), fp);
   sscanf(line, "%d", &numDesigns);

   // Process each design
   for (designNum = 1; designNum <= numDesigns; designNum++)
   {
      // Read in the design, first getting the number of junctions
      fgets(line, sizeof(line), fp);
      sscanf(line, "%d", &numJunctions);
      for (i = 0; i < numJunctions; i++)
      {
         // Get the next line of resistance values
         fgets(line, sizeof(line), fp);

         // Read the first resistance value from this line
         token = strtok(line, " \n");
         junctions[i][0] = atoi(token);

         // Read the rest of the resistances off of this line
         for (j = 1; j < numJunctions; j++)
         {
            token = strtok(NULL, " \n");
            junctions[i][j] = atoi(token);
         }
      }

      // Compute all paths of least resistance for this design, using
      // Floyd-Warshall's algorithm
      floyd();

      // Now, check the resistance on the path from the first junction
      // to the second, and print the results
      if (junctions[0][1] == 0)
         printf("Circuit Design #%d: Back to the drawing board\n", designNum);
      else
         printf("Circuit Design #%d: No more hedgehog troubles\n", designNum);
   }
}

File: UCF HSPT Project/src/y2007/roboto.java
import java.io.*;		// Import input and output libraries
import java.util.*;	// Import utility libraries (for Scanner)


class roboto
{

	// Create "infinity" constant to be a very large number, with regard to the 
	// problem. The maximum junction resistance in this problem is 500, so one 
	// billion effectivly acts like infinity
	public static final int INFINITY = 1000000000;

	/****************************************************************************/
	
	public static void main(String[] args) throws IOException
	{
		// Create input file scanner
		Scanner inpt = new Scanner(new File("roboto.in"));
		
		// Get the number of circuits to be tested
		int numDesigns = inpt.nextInt();
		
		for (int i = 0; i < numDesigns; i++)
		{
			// Get the number of junctions in this design
			int numJunctions = inpt.nextInt(); 
			
			// Create a 2D array to hold the junction connection information
			int[][] junctions = new int[numJunctions][numJunctions];
			
			// Get each junction connection resistance
			for (int j = 0; j < numJunctions; j++)
				for (int k = 0; k < numJunctions; k++)
				{
					junctions[j][k] = inpt.nextInt();

					// If there is no connection between j and k, as denoted by -1, there is 
					// essentially an infinite resistance between these junctions, so we set
					// the value of the resistance here to infinity
					if (junctions[j][k] == -1)
						junctions[j][k] = INFINITY;				
				}
				
			// Find the shortest paths between each junction using Floyd's Algorithm
			// (see function for more details)
			int[][] pathLengths = floyd(junctions);
			
			// If the shorest path from junction 0 to junction 1 is zero, then we know
			// that there is some path in the circuit that contains a zero resistance 
			// path between the two terminals, and thus there is a short circuit
			if (pathLengths[0][1] == 0)
				System.out.println("Circuit Design #" + 
				                   (i+1) + ": " +
				                   "Back to the drawing board");
			else
				System.out.println("Circuit Design #" + 
				                   (i+1) + ": " +
				                   "No more hedgehog troubles");
		}
		
		// All done now, so we can close the input file scanner
		inpt.close();
	}
	
	
	/****************************************************************************/
	
	/* 
		Floyd's Algorithm- finds the shortest path from each junction to each other 
	   junction. It does this by trying to find a shorter path from i to j by
		going through junction k. More information can be found at 
		http://en.wikipedia.org/wiki/Floyd%27s_algorithm
		
		The returned matrix is such that adj[i][j] = the length of the shortest path
		from i to j
	*/
	
	public static int[][] floyd (int[][] junctions)
	{
		//Make a copy becuase the matrix is destroyed through Floyd's
		int[][] adj = junctions.clone(); 
				
		//Start finding shortest paths		
		for (int k = 0; k < adj.length; k++)
			for (int i = 0; i < adj.length; i++)
				for (int j = 0; j < adj.length; j++)
					if (adj[i][k] + adj[k][j] < adj[i][j])
						adj[i][j] = adj[i][k] + adj[k][j];	
		
		return adj;		
	}
}
